src/addon/NUOPC/src/NUOPC_Driver.F90

```diff
@@ -994,113 +994,120 @@ recursive subroutine InitializeIPDv02p1(driver, importState, exportState, &
           line=__LINE__, file=trim(name)//":"//FILENAME)) return  ! bail out
 
         ! potentially the connector must be created here to/from driver-self
-        call NUOPC_CompAttributeGet(driver, name="HierarchyProtocol", &
-          isSet=needConnector, rc=rc)
+        call NUOPC_CompAttributeGet(driver, name="HierarchyConnectors", &
+          value=hierarchyConnectors, rc=rc)
         if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
           line=__LINE__, file=trim(name)//":"//FILENAME)) return  ! bail out
-        needConnector=.not.needConnector  ! default to do connection
-        if (.not.needConnector) then
-          ! inspect the HierarchyProtocol attribute to see if it requests a
-          ! connection
+        if(trim(hierarchyConnectors)=="auto") then
+          ! continue with automatic hierarchy connector creation here ...
           call NUOPC_CompAttributeGet(driver, name="HierarchyProtocol", &
-            value=hierarchyProtocol, rc=rc)
+            isSet=needConnector, rc=rc)
           if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
             line=__LINE__, file=trim(name)//":"//FILENAME)) return  ! bail out
-          if (trim(hierarchyProtocol)=="PushUpAllExportsAndUnsatisfiedImports" &
-            .or. trim(hierarchyProtocol)=="ConnectProvidedFields" &
-            .or. trim(hierarchyProtocol)=="Explorer") then
-            needConnector = .true.
-          endif
-        endif
-        areServicesSet = NUOPC_CompAreServicesSet(connector, rc=rc)
-        if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
-          line=__LINE__, file=trim(name)//":"//FILENAME)) return  ! bail out
-        if (.not.areServicesSet.and.(i==0.or.j==0)) then
-          ! the connector was not added by the user level code SetModelServices
-          ! and this involves the driver itself -> maybe automatic connector add
-          if (.not.(i==0.and.j==0)) then
-            ! not a driver-to-driver-self connection, which has no known purpose
-            if (i==0) then
-              needConnector = needConnector.and. &
-                ESMF_StateIsCreated(is%wrap%modelIS(i), rc=rc)
-              if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
-                line=__LINE__, file=trim(name)//":"//FILENAME)) return
-            else
-              ! j==0
-              needConnector = needConnector.and. &
-                ESMF_StateIsCreated(is%wrap%modelES(j), rc=rc)
-              if (ESMF_LogFoundError(rcToCheck=rc, msg=ESMF_LOGERR_PASSTHRU, &
-                line=__LINE__, file=trim(name)//":"//FILENAME)) return
+          needConnector=.not.needConnector  ! default to do connection
```

Comment:
This is really hard to follow - If a value is set for HierarchyProtocol then needConnector is false (because it's flipped). And if needConnector is false then condition is satisfied. Can you add a second boolean such as `hrchProtSet` then `needConnector=.not.hrchProtSet`.

Then change the next line to `if (hrchProtSet) then`

Reply:
Yep, I too had to look at the handling of `needConnector` for a bit before I understood it again when I came back to this code for this PR. However, I liked it again the way it was a after I understood it again.... but I also see your point.... so I just simplified it with an extra logical in 9eaf4ff.